layers/+window-management/exwm/funcs.el

```diff
@@ -78,6 +78,59 @@ Can show completions at point for COMMAND using helm"
   (interactive)
   (start-process "" nil exwm-locking-command))
 
+(defun exwm/autostart-process (name command &optional directory)
+  "Start a program is subprocess and append it in `exwm-autostart-process-list'.
+NAME is name for process.
+COMMAND is the shell command to run.
+DIRECTORY is the working directory in which the process is run.  It defaults
+  `user-home-directory' if not provided.
+The started process is also added to `exwm-autostart-process-list'."
+  (add-to-list (let ((default-directory (or directory user-home-directory)))
+                 (start-process-shell-command name nil command))
+               exwm-autostart-process-list))
+
+(defun exwm//start-desktop-application (xdg)
+  "Autostart an application from a XDG desktop entry specification."
+  (when-let* ((type (gethash "Type" xdg))
+              (_application? (string= type "Application"))
+              ;; Note that we can't always assume the presence of Exec: when
+              ;; DBusActivatable support is added, the application is launched
+              ;; through D-Bus rather than a program.
+              (basename (gethash "Name" xdg))
+              (cmd (gethash "Exec" xdg))
+                            ;; (dbus-p (gethash "DBusActivatable" xdg)) ; TODO: support
+              (directory (gethash "Path" xdg) user-home-directory)
+              (_include? (and (null (gethash "Hidden" xdg))
+                              (if-let ((only-show (gethash "OnlyShowIn" xdg)))
+                                  (member "EXWM" (split-string only-show ";" t))
+                                t)
+                              (if-let ((not-show (gethash "NotShowIn" xdg)))
+                                  (not (member "EXWM" (split-string not-show ";" t)))
+                                t)
+                              (if-let ((try-exec (gethash "TryExec" xdg)))
+                                  (executable-find try-exec)
+                                t))))
+    (exwm/autostart-process basename cmd directory)))
+
+(defun exwm//read-xdg-autostart-files ()
+  "Return a list for autostart applications as defined by the freedesktop specification."
+  (cl-loop with xdg-specs = (make-hash-table :test 'equal)
+           for dir in (append (xdg-config-dirs) (list (xdg-config-home)))
+           for autostart-dir = (concat dir "/autostart")
+           for autostart-files = (when (file-exists-p autostart-dir)
+                                   (directory-files autostart-dir t ".+\\.desktop$"))
+           do
+           (cl-loop for file in autostart-files do
```

Comment:
Instead of a hash table of hash table, we can make xdg-specs a list of hash table.

Instead of a nested `cl-loop`,
First, `for autostart-specs = (mapcar #'xdg-desktop-read-file autostart-files)`, you get a list of xdg-specs (for a given dir)

Next, you need to concat all xdg-specs for all dir, so you can use
`nconc xdg-specs` (no need to use the finally clause).

This returns a list of hash table

---

If you still believe xdg-specs needs to be a hash table of hash table, you cannot
`(push (xdg-desktop-read-file file) xdg-specs)`. Use `puthash`.

Reply:
Oops.  I was trying to make `xdg-specs` a list of hashtable, but I didn't change the definition from `make-hash-table` to `list`.